pytest/lib/mocknet.py

```diff
@@ -1193,59 +1245,83 @@ def start_genesis_updater(node, script, genesis_filename_in,
                           epoch_length, node_pks, increasing_stakes, num_seats,
                           single_shard, all_node_pks, node_ips, neard):
     logger.info(f'Starting genesis_updater on {node.instance_name}')
-    node.machine.run('bash',
-                     input=start_genesis_updater_script(
-                         script, genesis_filename_in, records_filename_in,
-                         config_filename_in, out_dir, chain_id, validator_keys,
-                         rpc_nodes, done_filename, epoch_length, node_pks,
-                         increasing_stakes, num_seats, single_shard,
-                         all_node_pks, node_ips, neard))
+    node.machine.run(
+        'bash',
+        input=start_genesis_updater_script(
+            script,
+            genesis_filename_in,
+            records_filename_in,
+            config_filename_in,
+            out_dir,
+            chain_id,
+            validator_keys,
+            rpc_nodes,
+            done_filename,
+            epoch_length,
+            node_pks,
+            increasing_stakes,
+            num_seats,
+            single_shard,
+            all_node_pks,
+            node_ips,
+            neard,
+        ),
+    )
 
 
-def start_genesis_update_waiter_script(done_filename):
-    return '''
-        rm /home/ubuntu/waiter.txt
-        until [ -f {done_filename} ]
-        do
-            echo 'waiting for {done_filename}' >> /home/ubuntu/waiter.txt
-            date >> /home/ubuntu/waiter.txt
-            ls {done_filename} >> /home/ubuntu/waiter.txt
-            sleep 5
-        done
-        echo 'File found' >> /home/ubuntu/waiter.txt
-    '''.format(dir=shlex.quote(PYTHON_DIR),
-               done_filename=shlex.quote(done_filename))
+def wait_genesis_updater_done(node, done_filename):
+    msg = f'Waiting for the genesis updater on {node.instance_name}'
+    logger.info(msg)
 
+    # Wait until the neard process stops running.
+    while True:
+        time.sleep(5)
+        result = node.machine.run('ps | grep neard')
```

Comment:
ah but what if `neard_amend_genesis_path()` returned None? Then `neard` will not be running, as it'll be a python script that updates the genesis files. Although honestly at this point, maybe we can just assume that `/home/ubuntu/neard.upgrade` will always have the `amend-genesis` command, since it was added a while ago now. So this part would be correct if you just remove `neard_amend_genesis_path()` entirely and just hardcode that we're going to assume that `/home/ubuntu/neard.upgrade` has the `amend-genesis` cmd. You could do that like:

```diff
diff --git a/pytest/lib/mocknet.py b/pytest/lib/mocknet.py
index c7bea792a..c43c0eeab 100644
--- a/pytest/lib/mocknet.py
+++ b/pytest/lib/mocknet.py
@@ -835,21 +835,11 @@ def create_genesis_file(validator_keys,
                         node_pks=None,
                         increasing_stakes=0.0,
                         num_seats=None,
-                        single_shard=False,
-                        neard=None):
-    if append and neard is not None:
-        neard_amend_genesis(neard, validator_keys, genesis_filename_in,
-                            records_filename_in, out_dir, rpc_node_names,
-                            chain_id, epoch_length, node_pks, increasing_stakes,
-                            num_seats, single_shard)
-    else:
-        genesis_filename_out = os.path.join(out_dir, 'genesis.json')
-        records_filename_out = os.path.join(out_dir, 'records.json')
-        do_create_genesis_file(validator_keys, genesis_filename_in,
-                               genesis_filename_out, records_filename_in,
-                               records_filename_out, rpc_node_names, chain_id,
-                               append, epoch_length, node_pks,
-                               increasing_stakes, num_seats, single_shard)
+                        single_shard=False):
+    neard_amend_genesis('/home/ubuntu/neard.upgrade', validator_keys, genesis_filename_in,
+                        records_filename_in, out_dir, rpc_node_names,
+                        chain_id, epoch_length, node_pks, increasing_stakes,
+                        num_seats, single_shard)
 
 
 def create_and_upload_genesis_file_from_empty_genesis(validator_node_and_stakes,
```
And you could delete `do_create_genesis_file()` (and maybe refactor it after as you see fit). and then also we don't need to pass `neard` to the genesis updater script:

```diff
diff --git a/pytest/lib/mocknet.py b/pytest/lib/mocknet.py
index c7bea792a..849d93bf0 100644
--- a/pytest/lib/mocknet.py
+++ b/pytest/lib/mocknet.py
@@ -1207,7 +1207,7 @@ def start_genesis_updater_script(script, genesis_filename_in,
                                  out_dir, chain_id, validator_keys, rpc_nodes,
                                  done_filename, epoch_length, node_pks,
                                  increasing_stakes, num_seats, single_shard,
-                                 all_node_pks, node_ips, neard):
+                                 all_node_pks, node_ips):
     validators = ','.join(
         [f'{account_id}={key}' for (account_id, key) in validator_keys.items()])
     cmd = ' '.join([
@@ -1230,7 +1230,6 @@ def start_genesis_updater_script(script, genesis_filename_in,
             single_shard,
             ','.join(all_node_pks),
             ','.join(node_ips),
-            neard if neard is not None else 'None',
         ]
     ])
     return '''
diff --git a/pytest/tests/mocknet/helpers/genesis_updater.py b/pytest/tests/mocknet/helpers/genesis_updater.py
index 8ce540300..ffeb77e25 100755
--- a/pytest/tests/mocknet/helpers/genesis_updater.py
+++ b/pytest/tests/mocknet/helpers/genesis_updater.py
@@ -51,10 +51,6 @@ def main(argv):
     node_ips = None
     if argv[15]:
         node_ips = argv[15].split(',')
-    if argv[16].lower() == 'none':
-        neard = None
-    else:
-        neard = argv[16]
 
     assert genesis_filename_in
     assert records_filename_in
@@ -81,8 +77,7 @@ def main(argv):
                                 node_pks=node_pks,
                                 increasing_stakes=increasing_stakes,
                                 num_seats=num_seats,
-                                single_shard=single_shard,
-                                neard=neard)
+                                single_shard=single_shard)
     config_filename_out = os.path.join(out_dir, 'config.json')
     mocknet.update_config_file(config_filename_in, config_filename_out,
                                all_node_pks, node_ips)
```

That is not a complete diff, since I didnt update the callers and whatnot, but that would be a sketch

Reply:
That sounds good but there is the 'append' arg, do you know what is that for? It seems that when append is false the python script should be used.